databases/core.py

```diff
@@ -63,8 +62,8 @@ def __init__(
         assert issubclass(backend_cls, DatabaseBackend)
         self._backend = backend_cls(self.url, **self.options)
 
-        # Connections are stored as task-local state.
-        self._connection_context: ContextVar = ContextVar("connection_context")
+        # Connections are stored per asyncio task
+        self._connections: typing.Dict[typing.Optional[asyncio.Task], Connection] = {}
```

Comment:
I think we can either exclude the `Optional` here _or_ exclude the assertions that `current_task` is not null. I don't have a strong preference. I omitted the assertions in my commit because they seemed onerous and `None` is perfectly fine to store in the dictionary. However, the assertion may lead to clearer behavior if someone tries to use this with a non-asyncio event loop — does this project support trio? If so we'll need to abstract current task retrieval to behave correctly.

Reply:
Yeah.. Looking into it, dropping optional would be the most correct right now. I was going by the return type annotation, but without providing a loop argument to `asyncio.current_task(...)`, this function will [use the loop actively running the task we're calling from](https://github.com/python/cpython/blob/1dad3d10714fe9611769196fde4e94bb63eb7f17/Lib/asyncio/tasks.py#L35-L39) or raise `RuntimeError: no running event loop`.

In other words, it appears that `asyncio.current_task(...)` can only return `None` if a loop argument is provided. Since we're never giving it a loop, the only thing we can get back is a `RuntimeError` or a `Task`. That's kind of unfortunate because the [type signature](https://github.com/python/typeshed/blob/0b17fafbe3a3a75ae11431ebdee72af5bf751a9c/stdlib/asyncio/tasks.pyi#L325)  doesn't have the overload to say that a return type of `None` is only possible if a loop is provided, so we still have to either ignore the Optional or assert / cast it away. 

I'm going to keep the asserts and drop the `Optional` type.